File: elixir/isogram/lib/isogram.ex
defmodule Isogram do
  @doc """
  Determines if a word or sentence is an isogram
  """
  @spec isogram?(String.t()) :: boolean
  def isogram?(sentence) do
    sentence
    |> String.replace(~r/[^a-zA-z]/, "")
    |> String.downcase()
    |> String.split("")
    |> Enum.reject(fn c -> c == "" end)
    |> Enum.reduce(%{}, fn x, occurances ->
      Map.update(occurances, x, 1, fn v -> v + 1 end)
    end)
    |> Map.values()
    |> Enum.all?(fn v -> v < 2 end)
  end
end

File: javascript/isogram/isogram.js
/**
 * 
 * @param {object} appearences 
 * @param {string} c 
 * @returns {object}
 */
function registerAppearence(appearences, c) {
  appearences[c] = (appearences[c] ?? 0) + 1;
  return appearences;
};

/**
 * 
 * @param {number} v 
 * @returns {boolean}
 */
function isUnique(v) {
  return v <= 1;
}

/**
 * 
 * @param {string} sentence
 * @returns {boolean}
 */
export const isIsogram = (sentence) => {
  return Object
    .values(
      sentence
        .replace(/[^a-zA-Z]/g, "")
        .toLowerCase()
        .split("")
        .reduce(registerAppearence, {})
    )
    .every(isUnique);
};
